doc/lexer_tutorial.md

```diff
@@ -0,0 +1,192 @@
+
+At the moment, LALRPOP doesn't allow you to configure the default tokenizer. In the future it will become quite flexible, but for now we have to write our own.
+
+Let's start by defining the stream format. The parser will accept an iterator where each item in the stream has the following structure:
+
+```rust
+pub type Spanned<Tok, Loc, Error> = Result<(Loc, Tok, Loc), Error>;
+```
+
+`Loc` is typically just a `usize`, representing a byte offset into the input string. Each token is accompanied by two of them, marking the start and end positions where it was found. `Error` can be pretty much anything you choose. And of course `Tok` is the meat of the stream, defining what possible values the tokens themselves can have. Following the conventions of Rust iterators, we'll signal a valid token with `Some(Ok(...))`, an error with `Some(Err(...))`, and EOF with `None`.
+
+(Note that the term "tokenizer" normally refers to a piece of code that simply splits up the stream, whereas a "lexer" also tags each token with its lexical category. What we're writing is the latter.)
+
+Whitespace is a simple language from a lexical standpoint, with only three valid tokens:
+
+```rust
+pub enum Tok {
+    Space,
+    Tab,
+    Linefeed,
+}
+```
+
+Everything else is a comment. There are no invalid lexes, so we'll define our own error type, a void enum:
+
+```rust
+pub enum LexicalError {
+    // Not possible
+}
+```
+
+Now for the lexer itself. We'll take a string slice as its input. For each token we process, we'll want to know the character value, and the byte offset in the string where it begins. We can do that by wrapping the `CharIndices` iterator, which yields tuples of `(usize, char)` representing exactly that information.
+
+```rust
+use std::str::CharIndices;
+
+pub struct Lexer<'input> {
+    chars: CharIndices<'input>,
+}
+
+impl<'input> Lexer<'input> {
+    pub fn new(input: &'input str) -> Self {
+        Lexer { chars: input.char_indices() }
+    }
+}
+```
+
+(The lifetime parameter `'input` indicates that the Lexer cannot outlive the string it's trying to parse.)
+
+Let's review our rules:
+
+- For a space character, we output `Tok::Space`.
+- For a tab character, we output `Tok::Tab`.
+- For a linefeed (newline) character, we output `Tok::Linefeed`.
+- We skip all other characters.
+- If we've reached the end of the string, we'll return `None` to signal EOF.
+
+Writing a lexer for a language with multi-character tokens can get very complicated, but this is so straightforward, we can translate it directly into code without thinking very hard. Here's our `Iterator` implementation:
+
+```rust
+impl<'input> Iterator for Lexer<'input> {
+    type Item = Spanned<Tok, usize, LexicalError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            return Some(Ok(match self.chars.next() {
```

Comment:
I wonder if this code would be clearer with the `return` moved into the arms? This structure is more concise but I fear it may throw people a bit. e.g.,

```rust
loop {
    match self.chars.next() {
        Some((i, ' ')) => return Some(Ok((i, Tok::Space, i+1))),
        ...
    }
}
```